aimmo-game/simulation/action.py

```diff
@@ -1,75 +1,154 @@
+    def __init__(self, avatar):
+        super(WaitAction, self).__init__(avatar)
+
+    def is_legal(self, world_map):
+        return True
+
+    def apply(self, world_map):
+        self.avatar.clear_action()
 
 
 class MoveAction(Action):
-    def __init__(self, direction):
+    def __init__(self, avatar, direction):
         # Untrusted data!
         self.direction = Direction(**direction)
+        super(MoveAction, self).__init__(avatar)
 
-    def apply(self, game_state, avatar):
-        target_location = avatar.location + self.direction
-        if game_state.world_map.can_move_to(target_location):
-            avatar.add_event(MovedEvent(avatar.location, target_location))
-            game_state.world_map.get_cell(avatar.location).avatar = None
-            avatar.location = target_location
-            new_cell = game_state.world_map.get_cell(target_location)
-            new_cell.avatar = avatar
-            if new_cell.pickup:
-                # TODO: potentially extract pickup logic into pickup when adding multiple types
-                avatar.health = min(10, avatar.health + new_cell.pickup.health_restored)
-                new_cell.pickup = None
-        else:
-            avatar.add_event(FailedMoveEvent(avatar.location, target_location))
-        _add_score_from_cell_if_needed(avatar, game_state)
+    def is_legal(self, world_map):
+        return world_map.can_move_to(self.target_location)
+
+    def process(self, world_map):
+        self.chain(world_map, {self.avatar.location})
+
+    def apply(self, world_map):
+        event = MovedEvent(self.avatar.location, self.target_location)
+        self.avatar.add_event(event)
+
+        world_map.get_cell(self.avatar.location).avatar = None
+        self.avatar.location = self.target_location
+        world_map.get_cell(self.target_location).avatar = self.avatar
+        self.avatar.clear_action()
+
+        new_cell = world_map.get_cell(self.target_location)
+        if new_cell.pickup:
+            # TODO:  extract pickup logic into pickup when adding multiple types
+            self.avatar.health = min(10, self.avatar.health + new_cell.pickup.health_restored)
+            new_cell.pickup = None
+
+        return True
+
+    def chain(self, world_map, visited):
+        if not self.is_legal(world_map):
+            return self.reject()
+
+        # Detect cycles
+        if self.target_location in visited:
+            return self.reject()
+
+        next_cell = world_map.get_cell(self.target_location)
+        if not next_cell.is_occupied:
+            return self.apply(world_map)
+
+        next_action = next_cell.avatar.action
+        if next_action.chain(world_map, visited | {self.target_location}):
+            return self.apply(world_map)
+
+        return self.reject()
+
+    def reject(self):
+        event = FailedMoveEvent(self.avatar.location, self.target_location)
+        self.avatar.add_event(event)
+        self.avatar.clear_action()
+        return False
```

Comment:
Filthy, filthy hack to avoid having to type `self.reject() ; return False` a bunch of times in `MoveAction.chain` below.